Assertion failed: class_decl && "TemplateSpecializationType is not a TplSpecDecl?"

Originally reported on Google Code with ID 150

```
I'm trying to use iwyu with http://github.com/mapbox/mapbox-gl-native on OS X 10.9.
I compiled iwyu against a clang 3.4.2 final stable.

I'm running this command:

CXX="include-what-you-use -Xiwyu --verbose=3 -isystem\"`xcode-select -p`/Toolchains/XcodeDefault.xctoolchain/usr/lib/c++/v1\""
make linux

and the first command gyp generates is

  include-what-you-use -Xiwyu --verbose=3 -isystem"/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/c++/v1"
'-DNDEBUG' -I../../include -I../Release/obj/gen/include  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk
-fasm-blocks -mpascal-strings -O3 -mmacosx-version-min=10.9 -arch x86_64 -std=c++11
-stdlib=libc++ -fvisibility-inlines-hidden -Wall -Wextra -I/usr/local/Cellar/libpng/1.6.10/include/libpng16
-I/usr/local/include -I/Users/kkaefer/Code/gl/native/mapnik-packaging/osx/out/build-cpp11-libcpp-universal/include
-MMD -MF ../Release/.deps/../Release/obj.target/llmr-x86/src/clipper/clipper.o.d.raw
 -c -o ../Release/obj.target/llmr-x86/src/clipper/clipper.o ../../src/clipper/clipper.cpp

which fails with

/Users/kkaefer/Code/include-what-you-use/iwyu.cc:3081: Assertion failed: class_decl
&& "TemplateSpecializationType is not a TplSpecDecl?"
make[1]: *** [../Release/obj.target/llmr-x86/src/clipper/clipper.o] Abort trap: 6

```

Reported by `mail@kkaefer.com` on 2014-06-17 14:38:03

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3306,9 +3306,18 @@ class InstantiatedTemplateVisitor
       return true;
     }
 
-    const ClassTemplateSpecializationDecl* class_decl
-        = DynCastFrom(TypeToDeclAsWritten(type));
-    CHECK_(class_decl && "TemplateSpecializationType is not a TplSpecDecl?");
+    const NamedDecl* named_decl = TypeToDeclAsWritten(type);
+    const ClassTemplateSpecializationDecl* class_decl = DynCastFrom(named_decl);
+
+    // Bail out if we are not a proper class
+    if (class_decl == nullptr) {
+      // If the template specialization decl is not sugar for a class, we
+      // expect it to be another kind of template decl, like a built-in.
+      CHECK_(llvm::isa<clang::TemplateDecl>(named_decl))
+          << "TemplateSpecializationType has no decl of type TemplateDecl?";
+      return true;
+    }
+
     if (ContainsKey(traversed_decls_, class_decl))
       return true;   // avoid recursion & repetition
     traversed_decls_.insert(class_decl);

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -216,7 +216,6 @@ using clang::ValueDecl;
 using clang::VarDecl;
 using llvm::cast;
 using llvm::dyn_cast;
-using llvm::dyn_cast_or_null;
 using llvm::errs;
 using llvm::isa;
 using std::make_pair;
@@ -2490,8 +2489,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (CanIgnoreCurrentASTNode())  return true;
     // ... only if this NewExpr involves a constructor call.
     const Expr* initializer = expr->getInitializer();
-    if (const CXXConstructExpr* cce =
-            dyn_cast_or_null<CXXConstructExpr>(initializer)) {
+    if (const CXXConstructExpr* cce = DynCastFrom(initializer)) {
       ReportIfReferenceVararg(cce->getArgs(),
                               cce->getNumArgs(),
                               cce->getConstructor());

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -1045,7 +1045,11 @@ class AstFlattenerVisitor : public BaseAstVisitor<AstFlattenerVisitor> {
     NodeSet* node_set = &nodeset_decl_cache_[decl];
     if (node_set->empty()) {
       if (decl->isImplicit()) {
-        TraverseImplicitDeclHelper(dyn_cast_or_null<FunctionDecl>(decl));
+        // TODO: For now, it is only working for functions. Check if it could
+        // make sense for other implicit decls too (e.g. BuiltinTemplateDecl)
+        if (FunctionDecl* func = DynCastFrom(decl)) {
+          TraverseImplicitDeclHelper(func);
+        }
       } else {
         TraverseDecl(decl);
       }

--- a/iwyu_ast_util.cc
+++ b/iwyu_ast_util.cc
@@ -120,7 +120,6 @@ using llvm::ArrayRef;
 using llvm::PointerUnion;
 using llvm::cast;
 using llvm::dyn_cast;
-using llvm::dyn_cast_or_null;
 using llvm::errs;
 using llvm::isa;
 using llvm::raw_string_ostream;
@@ -618,7 +617,7 @@ const RecordDecl* GetDefinitionForClass(const Decl* decl) {
   const RecordDecl* as_record = DynCastFrom(decl);
   const ClassTemplateDecl* as_tpl = DynCastFrom(decl);
   if (as_tpl) {  // Convert the template to its underlying class defn.
-    as_record = dyn_cast_or_null<RecordDecl>(as_tpl->getTemplatedDecl());
+    as_record = DynCastFrom(as_tpl->getTemplatedDecl());
   }
   if (as_record) {
     if (const RecordDecl* record_dfn = as_record->getDefinition()) {

--- a/iwyu_output.cc
+++ b/iwyu_output.cc
@@ -1120,16 +1120,27 @@ void ProcessFullUse(OneUse* use,
     }
   }
 
-  // (B3) Discard symbol uses for builtin symbols, including new/delete.
-  // TODO(csilvers): we could use getBuiltinID(), but it returns
-  // non-zero for things like malloc.  Figure out how to use it.
-  if (const FunctionDecl* fn_decl = DynCastFrom(use->decl())) {
-    if (StartsWith(use->symbol_name(), "__builtin_")) {
+  // (B3) Discard symbol uses for builtin symbols, including new/delete and
+  // template builtins.
+  if (isa<clang::BuiltinTemplateDecl>(use->decl())) {
+    VERRS(6) << "Ignoring use of " << use->symbol_name()
+             << " (" << use->PrintableUseLoc() << "): built-in template\n";
+    use->set_ignore_use();
+    return;
+  }
+  // A compiler builtin without a predefined header file (e.g. __builtin_..)
+  if (const clang::IdentifierInfo* iden = use->decl()->getIdentifier()) {
+    if (iden->getBuiltinID() != 0 &&
+        !clang::Builtin::Context::isBuiltinFunc(use->symbol_name().c_str())) {
       VERRS(6) << "Ignoring use of " << use->symbol_name()
                << " (" << use->PrintableUseLoc() << "): built-in function\n";
       use->set_ignore_use();
       return;
     }
+  }
+  // Special case for operators new/delete: Only treated as built-in if they
+  // are the default, non-placement versions.
+  if (const FunctionDecl* fn_decl = DynCastFrom(use->decl())) {
     const string dfn_file = GetFilePath(fn_decl);
     if (IsDefaultNewOrDelete(fn_decl, ConvertToQuotedInclude(dfn_file))) {
       VERRS(6) << "Ignoring use of " << use->symbol_name()

--- a/tests/cxx/built_ins_template.cc
+++ b/tests/cxx/built_ins_template.cc
@@ -1,0 +1,22 @@
+//===--- built_ins_template.cc - test input file for iwyu -----------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// Built-in templates have no declaration, thus they need no include.
+
+__type_pack_element<0, int> tp;
+
+template <class T, T...> struct A {};
+__make_integer_seq<A, int, 5> seq;
+
+
+/**** IWYU_SUMMARY
+
+(tests/cxx/built_ins_template.cc has correct #includes/fwd-decls)
+
+***** IWYU_SUMMARY */

